src/configSchema.ts

```diff
@@ -0,0 +1,167 @@
+ */
+
+function transformDurationString(durationStr: string, ctx) {
+	const duration = ms(durationStr);
+	if (isNaN(duration)) {
+		// adds the error to the Zod Issues
+		ctx.addIssue({
+			code: "custom",
+			message: zodErrorMsg.vercel,
+		});
+	}
+	return duration;
+}
+
+/**
+ * an object of the zod schema
+ * each are named after what they are intended to validate
+ */
+
+export const VALIDATION_SCHEMA = z
+	.object({
+		delay: z
+			.number()
+			.positive({
+				message: zodErrorMsg.delay,
+			})
+			.default(10),
+		torznab: z.array(z.string().url()),
+		dataDirs: z.array(z.string()).nullish(),
+		matchMode: z.nativeEnum(MatchMode),
+		dataCategory: z.string().nullish(),
+		linkDir: z.string().nullish(),
+		linkType: z.nativeEnum(LinkType),
+		skipRecheck: z.boolean(),
+		maxDataDepth: z.number().gte(1),
+		torrentDir: z.string(),
+		outputDir: z.string(),
+		includeEpisodes: z.boolean(),
+		includeSingleEpisodes: z.boolean(),
+		includeNonVideos: z.boolean(),
+		fuzzySizeThreshold: z.number().positive().lte(1, {
+			message: zodErrorMsg.fuzzySizeThreshold,
+		}),
+		excludeOlder: z
+			.string()
+			.min(1, { message: zodErrorMsg.emptyString })
+			.transform(transformDurationString)
+			.nullish(),
+		excludeRecentSearch: z
+			.string()
+			.min(1, { message: zodErrorMsg.emptyString })
+			.transform(transformDurationString)
+			.nullish(),
+		action: z.nativeEnum(Action),
+		qbittorrentUrl: z.string().url().nullish(),
+		rtorrentRpcUrl: z.string().url().nullish(),
+		transmissionRpcUrl: z.string().url().nullish(),
+		delugeRpcUrl: z.string().url().nullish(),
+		duplicateCategories: z.boolean(),
+		notificationWebhookUrl: z.string().url().nullish(),
+		port: z
+			.number()
+			.positive()
+			.lte(65535)
+			.or(z.boolean().transform(() => null)),
+		host: z.string().ip().nullish(),
+		rssCadence: z
+			.string()
+			.min(1, { message: zodErrorMsg.emptyString })
+			.transform(transformDurationString)
+			.nullish(),
+		searchCadence: z
+			.string()
+			.min(1, { message: zodErrorMsg.emptyString })
+			.transform(transformDurationString)
+			.nullish(),
+		snatchTimeout: z
+			.string()
+			.min(1, { message: zodErrorMsg.emptyString })
+			.transform(transformDurationString)
+			.nullish(),
+		searchTimeout: z
+			.string()
+			.min(1, { message: zodErrorMsg.emptyString })
+			.transform(transformDurationString)
+			.nullish(),
+		searchLimit: z.number().positive().nullish(),
+		apiAuth: z.boolean().default(false),
+		verbose: z.boolean().default(false),
+		torrents: z.array(z.string()).optional(),
+	})
+	.strict()
+	.refine(
+		(config) =>
+			config.action !== Action.INJECT ||
+			config.rtorrentRpcUrl ||
+			config.qbittorrentUrl ||
+			config.transmissionRpcUrl ||
+			config.delugeRpcUrl,
+		() => ({
+			message: zodErrorMsg.injectUrl,
+		})
+	)
+	.refine(
+		(config) => {
+			if (
+				(config.dataDirs !== undefined &&
+					config.dataDirs !== null &&
+					config.dataDirs?.length > 0) ||
```

Comment:
This actually shouldn't be particularly hard, there are a few approaches.

We can use `nonempty()` in conjunction with the array, and error on empty arrays altogether, or alternatively transform an empty array into undefined.

Either way, it isn't entirely outside the scope of this PR at all IMO, and either approach is pretty easy to implement. The latter would be my preference and would result in not breaking any existing dependence on it being undefined.

I'll try and get that implemented here in the next commit

Reply:
From my testing it seems like I've accounted for empty arrays now.

Reply:
So the empty array transforms to undefined now, eliminating the need to fallback to anything in commander. It also logs in configdump as `dataDirs: undefined` rather than omitting it as it would normally if undefined in the config, so I think this is pretty much the best solution I can come up with currently.

If you have other ideas on handling this, feel free to give me the input.

Reply:
This seems fine, but I guess - why would an empty array be bad?

Reply:
I don't know how it was handled downstream, but it seem logical to do it that way to me at least.

Reply:
OK after reading the code I think I'm less concerned. As long as we still accept empty arrays, we can use whatever representation we want internally.

Reply:
Great, if you want to change this in the future it's pretty inconsequential to transform empty arrays to whatever we deem useful.